File: src/tui/selection-state.ts
export type ParentSelectionState = "empty" | "partial" | "full";

export type TreeSelectionState = {
  allLeafIds: string[];
  selectedLeafIds: string[];
};

export function getParentSelectionState(
  state: TreeSelectionState,
): ParentSelectionState {
  if (state.selectedLeafIds.length === 0) {
    return "empty";
  }
  if (state.selectedLeafIds.length === state.allLeafIds.length) {
    return "full";
  }
  return "partial";
}

// parent
//   [ ] -> [x] -> [-] transitions come from child toggles
// children
//   toggle child recalculates parent from current selected set
export function toggleParent(state: TreeSelectionState): TreeSelectionState {
  return getParentSelectionState(state) === "full"
    ? { ...state, selectedLeafIds: [] }
    : { ...state, selectedLeafIds: [...state.allLeafIds] };
}

export function toggleChild(
  state: TreeSelectionState,
  leafId: string,
): TreeSelectionState {
  const selected = new Set(state.selectedLeafIds);
  if (selected.has(leafId)) {
    selected.delete(leafId);
  } else {
    selected.add(leafId);
  }

  return {
    ...state,
    selectedLeafIds: state.allLeafIds.filter((id) => selected.has(id)),
  };
}

File: apps/desktop-mac/Sources/DesktopApp/Store/SelectionState.swift
import Foundation

enum SelectionState: String, Equatable {
    case empty
    case partial
    case full
}

struct TreeSelectionState {
    var allLeafIds: [String]
    var selectedLeafIds: [String]
}

func selectionState(allIds: [String], selectedIds: [String]) -> SelectionState {
    getParentSelectionState(TreeSelectionState(allLeafIds: allIds, selectedLeafIds: selectedIds))
}

func getParentSelectionState(_ state: TreeSelectionState) -> SelectionState {
    if state.selectedLeafIds.isEmpty {
        return .empty
    }
    if state.selectedLeafIds.count == state.allLeafIds.count {
        return .full
    }
    return .partial
}

func toggleParent(_ state: TreeSelectionState) -> TreeSelectionState {
    getParentSelectionState(state) == .full
        ? TreeSelectionState(allLeafIds: state.allLeafIds, selectedLeafIds: [])
        : TreeSelectionState(allLeafIds: state.allLeafIds, selectedLeafIds: state.allLeafIds)
}

func toggleChild(_ state: TreeSelectionState, leafId: String) -> TreeSelectionState {
    let selected = Set(state.selectedLeafIds)
    var nextSelected = selected
    if nextSelected.contains(leafId) {
        nextSelected.remove(leafId)
    } else {
        nextSelected.insert(leafId)
    }

    return TreeSelectionState(
        allLeafIds: state.allLeafIds,
        selectedLeafIds: state.allLeafIds.filter { nextSelected.contains($0) }
    )
}
